client/signup/steps/p2-site/index.jsx

```diff
@@ -165,12 +183,45 @@ class P2Site extends React.Component {
 
 							this.logValidationErrorToLogstash( error.error, errorMessage );
 						} else {
-							messages.site = {
-								[ error.error ]: error.message,
-							};
-
+							if ( SITE_TAKEN_ERROR_CODES.includes( error.error ) ) {
+								messages.site = {
+									[ ERROR_CODE_TAKEN_SITE ]: this.props.translate(
+										'Sorry, that site already exists! Here are some available alternatives:'
+									),
+								};
+							} else {
+								messages.site = {
+									[ error.error ]: error.message,
+								};
+							}
+
+							// We want to log the real error code and message. The above is formatted for the end user
+							// only.
 							this.logValidationErrorToLogstash( error.error, error.message );
 						}
+
+						if ( error.error && SITE_TAKEN_ERROR_CODES.includes( error.error ) ) {
+							WPCOM_SUBDOMAIN_SUFFIX_SUGGESTIONS.forEach( ( suffix ) => {
+								const suggestedSubdomain = `${ fields.site }${ suffix }.wordpress.com`;
+
+								wpcom
+									.domains()
+									.suggestions( {
+										quantity: 1,
+										query: suggestedSubdomain,
+										only_wordpressdotcom: true,
```

Comment:
Just wanted to make this clear: you have two options for the algorithm to generate suggestions for WPCOM subdomains.

`include_wordpressdotcom = true` and `only_wordpressdotcom = false` will use local, crude suggestion that simply appends a random number to the suggestion if the queried one is not available. So `examplep2` will result in `examplep29570384023`. This one guarantees you'll have `examplep2` in the suggestion. But the suggestion itself might not be attractive in many cases (for generic queries).

`only_wordpressdotcom = true` (`include_wordpressdotcom` defaults to `false`) will use a third-party which can result in nicer, less random suggestions. But it does not guarantee that `examplep2` will be part of the generated suggestion. It will, however, result in more human-friendly suggestions (...in general ;)).

For _example_:
<img width="388" alt="Screenshot 2020-08-28 at 16 25 25" src="https://user-images.githubusercontent.com/3392497/91590698-4654c100-e94b-11ea-9009-b6d75e48b66c.png">
(note: no `examplep2` result - instead there's a weird `examplec1` one)

vs

<img width="387" alt="Screenshot 2020-08-28 at 16 25 41" src="https://user-images.githubusercontent.com/3392497/91590706-49e84800-e94b-11ea-846d-936457761d90.png">

I'm guessing the current approach will suit this flow better, but wanted to make sure you understand the nuances before the two options 🙇

Reply:
Ah, good to know, thanks!

I think we are combining the best of both worlds here: querying the suggestions endpoint with `only_wordpressdotcom: true` and `quantity: 1` so we get the desired suffix in the wpcom domain if it's available and a nice suggestion if it's not. I guess getting the nice suggestions instead of our desired ones would only be a problem if we set the quantity to more than 1.